src/Features/CSharp/Portable/Snippets/CSharpConstructorSnippetProvider.cs

```diff
@@ -36,10 +36,9 @@ public CSharpConstructorSnippetProvider()
         {
         }
 
-        protected override async Task<bool> IsValidSnippetLocationAsync(Document document, int position, CancellationToken cancellationToken)
+        protected override bool IsValidSnippetLocation(SyntaxContext context, CancellationToken cancellationToken)
```

Comment:
can you make this inheritance hierarchy generic on TSyntaxContext?  you wouldn't need the cast below then.

Reply:
I might be missing something obvious.

But `GetSnippetProviders` needs to cast providers from `ISnippetProvider` to `AbstractSnippetProvider`. If `AbstractSnippetProvider` is now generic, things will get more complicated and trickier.

I guess the cast is okay, we already do similar thing in analyzers where we cast a `Compilation` to `CSharpCompilation` for example?

Also the situation will get clearer once VB snippets are implemented. For example, we can see how the VB Constructor snippet will implement `IsValidSnippetLocation` and see if we can make a common abstract method in the base abstract `SyntaxContext`.